File: src/electron/function/windowBounds.ts
export const LEGACY_DEFAULT_WINDOW_SIZE = { width: 850, height: 530 } as const
export const IDEAL_WINDOW_SIZE = { width: 1200, height: 800 } as const
export const PREFERRED_MIN_WINDOW_SIZE = { width: 720, height: 450 } as const
export const HARD_MIN_WINDOW_SIZE = { width: 350, height: 450 } as const

const WORK_AREA_MARGIN = 48
const WIDTH_RATIO = 0.72
const HEIGHT_RATIO = 0.78
const MIN_DEFAULT_WIDTH = 960
const MIN_DEFAULT_HEIGHT = 640
const LEGACY_SIZE_TOLERANCE = 8

export interface WorkArea {
    x: number
    y: number
    width: number
    height: number
}

export interface WindowBounds {
    x: number
    y: number
    width: number
    height: number
}

export interface RestoredWindowBounds {
    x?: number
    y?: number
    width: number
    height: number
}

function clamp(value: number, min: number, max: number): number {
    if (max < min) return max
    return Math.min(max, Math.max(min, value))
}

export function isLegacyDefaultWindowSize(size: { width: number; height: number }): boolean {
    return Math.abs(size.width - LEGACY_DEFAULT_WINDOW_SIZE.width) <= LEGACY_SIZE_TOLERANCE
        && Math.abs(size.height - LEGACY_DEFAULT_WINDOW_SIZE.height) <= LEGACY_SIZE_TOLERANCE
}

export function minWindowSize(workArea: { width: number; height: number }): { width: number; height: number } {
    return {
        width: Math.max(HARD_MIN_WINDOW_SIZE.width, Math.min(PREFERRED_MIN_WINDOW_SIZE.width, Math.floor(workArea.width))),
        height: Math.max(HARD_MIN_WINDOW_SIZE.height, Math.min(PREFERRED_MIN_WINDOW_SIZE.height, Math.floor(workArea.height))),
    }
}

export function computeDefaultWindowSize(workArea: { width: number; height: number }): { width: number; height: number } {
    const maxWidth = Math.max(HARD_MIN_WINDOW_SIZE.width, workArea.width - WORK_AREA_MARGIN)
    const maxHeight = Math.max(HARD_MIN_WINDOW_SIZE.height, workArea.height - WORK_AREA_MARGIN)
    const width = clamp(
        Math.round(Math.min(IDEAL_WINDOW_SIZE.width, workArea.width * WIDTH_RATIO)),
        Math.min(MIN_DEFAULT_WIDTH, maxWidth),
        maxWidth,
    )
    const height = clamp(
        Math.round(Math.min(IDEAL_WINDOW_SIZE.height, workArea.height * HEIGHT_RATIO)),
        Math.min(MIN_DEFAULT_HEIGHT, maxHeight),
        maxHeight,
    )
    return { width, height }
}

export function centerWindowOnWorkArea(
    size: { width: number; height: number },
    workArea: WorkArea,
): WindowBounds {
    return {
        x: Math.round(workArea.x + (workArea.width - size.width) / 2),
        y: Math.round(workArea.y + (workArea.height - size.height) / 2),
        width: size.width,
        height: size.height,
    }
}

export function fitWindowToWorkArea(bounds: RestoredWindowBounds, workArea: WorkArea): WindowBounds {
    const minSize = minWindowSize(workArea)
    const maxWidth = Math.max(minSize.width, workArea.width - WORK_AREA_MARGIN)
    const maxHeight = Math.max(minSize.height, workArea.height - WORK_AREA_MARGIN)
    const width = clamp(Math.round(bounds.width), minSize.width, maxWidth)
    const height = clamp(Math.round(bounds.height), minSize.height, maxHeight)
    if (bounds.x === undefined || bounds.y === undefined || !Number.isFinite(bounds.x) || !Number.isFinite(bounds.y)) {
        return centerWindowOnWorkArea({ width, height }, workArea)
    }
    const x = clamp(Math.round(bounds.x), workArea.x, Math.max(workArea.x, workArea.x + workArea.width - width))
    const y = clamp(Math.round(bounds.y), workArea.y, Math.max(workArea.y, workArea.y + workArea.height - height))
    return { x, y, width, height }
}

export function resolveWindowBounds(restored: RestoredWindowBounds, workArea: WorkArea): WindowBounds {
    const legacy = isLegacyDefaultWindowSize(restored)
    const size = legacy ? computeDefaultWindowSize(workArea) : restored
    const next: RestoredWindowBounds = {
        width: size.width,
        height: size.height,
    }
    if (!legacy && restored.x !== undefined) next.x = restored.x
    if (!legacy && restored.y !== undefined) next.y = restored.y
    return fitWindowToWorkArea(next, workArea)
}

File: src/tauri/src/window_bounds.rs
pub const LEGACY_WIDTH: f64 = 850.0;
pub const LEGACY_HEIGHT: f64 = 530.0;
pub const IDEAL_WIDTH: f64 = 1200.0;
pub const IDEAL_HEIGHT: f64 = 800.0;
pub const PREFERRED_MIN_WIDTH: f64 = 720.0;
pub const PREFERRED_MIN_HEIGHT: f64 = 450.0;
pub const HARD_MIN_WIDTH: f64 = 350.0;
pub const HARD_MIN_HEIGHT: f64 = 450.0;

const WORK_AREA_MARGIN: f64 = 48.0;
const WIDTH_RATIO: f64 = 0.72;
const HEIGHT_RATIO: f64 = 0.78;
const MIN_DEFAULT_WIDTH: f64 = 960.0;
const MIN_DEFAULT_HEIGHT: f64 = 640.0;
const LEGACY_SIZE_TOLERANCE: f64 = 8.0;

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct WindowBounds {
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
}

fn clamp(value: f64, min: f64, max: f64) -> f64 {
    if max < min {
        max
    } else {
        value.max(min).min(max)
    }
}

pub fn is_legacy_default_window_size(width: f64, height: f64) -> bool {
    (width - LEGACY_WIDTH).abs() <= LEGACY_SIZE_TOLERANCE
        && (height - LEGACY_HEIGHT).abs() <= LEGACY_SIZE_TOLERANCE
}

pub fn min_window_size(work_width: f64, work_height: f64) -> (f64, f64) {
    (
        HARD_MIN_WIDTH.max(PREFERRED_MIN_WIDTH.min(work_width.floor())),
        HARD_MIN_HEIGHT.max(PREFERRED_MIN_HEIGHT.min(work_height.floor())),
    )
}

pub fn compute_default_window_size(work_width: f64, work_height: f64) -> (f64, f64) {
    let max_width = HARD_MIN_WIDTH.max(work_width - WORK_AREA_MARGIN);
    let max_height = HARD_MIN_HEIGHT.max(work_height - WORK_AREA_MARGIN);
    let width = clamp(
        (IDEAL_WIDTH.min(work_width * WIDTH_RATIO)).round(),
        MIN_DEFAULT_WIDTH.min(max_width),
        max_width,
    );
    let height = clamp(
        (IDEAL_HEIGHT.min(work_height * HEIGHT_RATIO)).round(),
        MIN_DEFAULT_HEIGHT.min(max_height),
        max_height,
    );
    (width, height)
}

pub fn fit_window_to_work_area(
    x: Option<f64>,
    y: Option<f64>,
    width: f64,
    height: f64,
    work_x: f64,
    work_y: f64,
    work_width: f64,
    work_height: f64,
) -> WindowBounds {
    let (min_width, min_height) = min_window_size(work_width, work_height);
    let max_width = min_width.max(work_width - WORK_AREA_MARGIN);
    let max_height = min_height.max(work_height - WORK_AREA_MARGIN);
    let width = clamp(width.round(), min_width, max_width);
    let height = clamp(height.round(), min_height, max_height);
    match (x, y) {
        (Some(x), Some(y)) if x.is_finite() && y.is_finite() => WindowBounds {
            x: clamp(
                x.round(),
                work_x,
                work_x.max(work_x + work_width - width),
            ),
            y: clamp(
                y.round(),
                work_y,
                work_y.max(work_y + work_height - height),
            ),
            width,
            height,
        },
        _ => WindowBounds {
            x: (work_x + (work_width - width) / 2.0).round(),
            y: (work_y + (work_height - height) / 2.0).round(),
            width,
            height,
        },
    }
}

pub fn resolve_window_bounds(
    x: Option<f64>,
    y: Option<f64>,
    width: f64,
    height: f64,
    work_x: f64,
    work_y: f64,
    work_width: f64,
    work_height: f64,
) -> WindowBounds {
    let legacy = is_legacy_default_window_size(width, height);
    let (width, height) = if legacy {
        compute_default_window_size(work_width, work_height)
    } else {
        (width, height)
    };
    fit_window_to_work_area(
        if legacy { None } else { x },
        if legacy { None } else { y },
        width,
        height,
        work_x,
        work_y,
        work_width,
        work_height,
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn uses_ideal_size_on_1080p() {
        assert_eq!(compute_default_window_size(1920.0, 1080.0), (1200.0, 800.0));
    }

    #[test]
    fn replaces_legacy_default() {
        let bounds = resolve_window_bounds(
            Some(40.0),
            Some(60.0),
            850.0,
            530.0,
            0.0,
            0.0,
            1920.0,
            1080.0,
        );
        assert_eq!(
            bounds,
            WindowBounds {
                x: 360.0,
                y: 140.0,
                width: 1200.0,
                height: 800.0,
            }
        );
    }
}
